File: stdlib.c

#define gostdc_stdio_h
#define gostdc_time_h
#define gostdc_locale_h
#define gostdc_stdarg_h

#include "gostdc.h"


FILE* stderr;
FILE* stdin;
FILE* stdout;

void
·initlibs(void) {
    stdin = (FILE*)3;   // better not 0, because 0 means NULL means sometimes error
    stdout = (FILE*)1;
    stderr = (FILE*)2;
}

////////////////////

void runtime·panicstring(int8*);

static char* nyi_string = "0 not yet implemented";
void
nyi(char char_id) {
    nyi_string[0] = char_id;
    runtime·panicstring(nyi_string);
    //runtime·panicstring("not yet implemented");
}

void
exit(int code) {
    void ·go_exit(uintptr);
    ·go_exit(code);
}

void*
realloc(void* ptr, size_t size) {
    void ·go_realloc(uintptr, uintptr, uintptr);
    void* ret = 0;
    ·go_realloc((uintptr)ptr, (uintptr)size, (uintptr)&ret);
    return ret;
}

void
free(void* ptr) {
    void ·go_free(uintptr);
    ·go_free((uintptr)ptr);
}

double
pow(double x, double y) {
    void ·go_pow(uintptr, uintptr);
    ·go_pow((uintptr)&x, (uintptr)&y);
    return x;
}

double
fmod(double x, double y) {
    void ·go_fmod(uintptr, uintptr);
    ·go_fmod((uintptr)&x, (uintptr)&y);
    return x;
}

double
modf(double x, double *iptr) {
    void ·go_modf(uintptr, uintptr);
    ·go_modf((uintptr)&x, (uintptr)iptr);
    return x;
}

double
atan2(double y, double x) {
    void ·go_atan2(uintptr, uintptr);
    ·go_atan2((uintptr)&y, (uintptr)&x);
    return y;
}

//FIXME
#pragma textflag 7
int
sprintf(char* str, const char* format, ...) {
    void ·go_sprintf(uintptr, uintptr, uintptr, uintptr);
    va_list argp;
    va_start(argp, format);
    ·go_sprintf((uintptr)str, (uintptr)format, (uintptr)argp, (uintptr)_BIGWORD);
    va_end(argp);
    //nyi('6');
    return 0; //FIXME
}

//FIXME
#pragma textflag 7
int
fprintf(FILE* stream, const char* format, ...) {
    void ·go_fprintf(uintptr, uintptr, uintptr, uintptr);
    va_list argp;
    va_start(argp, format);
    ·go_fprintf((uintptr)stream, (uintptr)format, (uintptr)argp, (uintptr)_BIGWORD);
    va_end(argp);
    //nyi('7');
    return 0; //FIXME
}

//FIXME
int
fclose(FILE* stream) {
    nyi('8');
    return 0;
}

//FIXME
int
ferror(FILE* stream) {
    nyi('9');
    return 0;
}

//FIXME
FILE*
freopen(const char* filename, const char* mode, FILE* stream) {
    nyi('a');
    return stderr;
}

//FIXME
FILE*
fopen(const char* filename, const char* mode) {
    nyi('b');
    return stderr;
}

//FIXME
int
getc(FILE* stream) {
    nyi('c');
    return -1;
}

//FIXME
int
ungetc(int ch, FILE* stream) {
    nyi('d');
    return -1;
}

//FIXME
size_t
fread(void* ptr, size_t size, size_t count, FILE* stream) {
    nyi('e');
    return 0;
}

//FIXME
int
feof(FILE* stream) {
    nyi('f');
    return -1;
}

void ·goputc(uintptr, uintptr);

//FIXME
int
fputs(const char* s, FILE* stream) {
    while (*s) {
        ·goputc((uintptr)*s, (uintptr)stream);
        s++;
    }
    return 1; //FIXME
}

//FIXME
char*
strerror(int) {
    nyi('g');
    return "STRERROR";
}

//FIXME
double
strtod(const char* nptr, char** endptr) {
    void ·go_strtod(uintptr, uintptr, uintptr presult);
    double result=0;
    ·go_strtod((uintptr)nptr, (uintptr)endptr, (uintptr)&result);
    return result;
}

int errno;

//FIXME
int isxdigit(int c) {
    return (c>='0' && c<='9') ||
        (c>='a' && c<='f') ||
        (c>='A' && c<='F');
}

//FIXME
int isupper(int c) {
    return (c>='A' && c<='Z');
}

//FIXME
int islower(int c) {
    return (c>='a' && c<='z');
}

//FIXME
int isdigit(int c) {
    return (c>='0' && c<='9');
}

//FIXME
int isalpha(int c) {
    return isupper(c) || islower(c);
}

//FIXME
int isalnum(int c) {
    return isalpha(c) || isdigit(c);
}

//FIXME
int iscntrl(int c) {
    return (c>=0 && c<=0x1f) || c==0x7f;
}

//FIXME
int
isspace(int c) {
    return (c == ' ') ||
        (c == '\n') ||
        (c == '\r') ||
        (c == '\t');
}

//FIXME
int ispunct(int c) {
    return c=='!' ||
        c=='\"' ||
        c=='#' ||
        c=='$' ||
        c=='%' ||
        c=='&' ||
        c=='\'' ||
        c=='(' ||
        c==')' ||
        c=='*' ||
        c=='+' ||
        c==',' ||
        c=='-' ||
        c=='.' ||
        c=='/' ||
        c==':' ||
        c==';' ||
        c=='<' ||
        c=='=' ||
        c=='>' ||
        c=='?' ||
        c=='@' ||
        c=='[' ||
        c=='\\' ||
        c==']' ||
        c=='^' ||
        c=='_' ||
        c=='`' ||
        c=='{' ||
        c=='|' ||
        c=='}' ||
        c=='~';
}

int toupper(int c) {
    if (c>='a' && c<='z') {
        return c - 'a' + 'A';
    }
    return c;
}

int tolower(int c) {
    if (c>='A' && c<='Z') {
        return c - 'A' + 'a';
    }
    return c;
}

//FIXME
int abs(int x) {
    if (x<0)
        return -x;
    return x;
}

//FIXME
double
frexp(double num, int* exp) {
    *exp=0;
    return 0;
/*
    void ·go_frexp(uintptr, uintptr);
    ·go_frexp((uintptr)&num, (uintptr)exp); //FIXME: is int* safe here, with regards to its width?
    return num;
*/
}

//FIXME
double
ldexp(double x, int exp) {
    return 0;
/*
    void ·go_ldexp(uintptr, uintptr);
    ·go_ldexp((uintptr)&x, (uintptr)&exp); //FIXME: is int* safe here, with regards to its width?
    return x;
*/
}

int
rand(void) {
    void ·go_rand(uintptr);
    int32 ret = 0;
    ·go_rand((uintptr)&ret);
    return ret;
}

void
srand(unsigned seed) {
    void ·go_srand(uintptr);
    ·go_srand((uintptr)seed);
}

File: stdlib.go
package gostdc

import (
	"fmt"
	"math"
	"math/rand"
	"os"
	"runtime"
	"strconv"
	"testing"
	"time"
	"unsafe"
)

func peek(addr uintptr) *byte {
	return (*byte)(unsafe.Pointer(addr))
}

type alloc struct {
	p unsafe.Pointer
	s []byte
}

var allocs = make(map[uintptr]alloc)

func go_exit(code int) {
	os.Exit(code)
}

func go_realloc(ptr uintptr, size uint32, pnewptr uintptr) {
	var newptr uintptr = 0
	switch {
	case ptr == 0 && size == 0:
		break
	case ptr == 0: // malloc
		//TODO: if (size > int32.MAX) return 0; and set errno?
		if int(size) < 0 {
			break
		}
		buf := alloc{s: make([]byte, int(size))}
		buf.p = unsafe.Pointer(&buf.s[0])
		allocs[uintptr(buf.p)] = buf
		newptr = uintptr(buf.p)
	case size == 0: // free
		delete(allocs, ptr)
		break
	default: // realloc
		tmp := allocs[ptr].s
		go_realloc(ptr, 0, uintptr(unsafe.Pointer(&newptr)))
		go_realloc(0, size, uintptr(unsafe.Pointer(&newptr)))
		copy(allocs[newptr].s, tmp)
	}
	*((*uintptr)(unsafe.Pointer(pnewptr))) = newptr
}

func go_free(ptr uintptr) {
	go_realloc(ptr, 0, uintptr(unsafe.Pointer(&ptr)))
}

func go_pow(x, y uintptr) {
	px := (*float64)(unsafe.Pointer(x))
	py := (*float64)(unsafe.Pointer(y))
	*px = math.Pow(*px, *py)
}

func go_fmod(x, y uintptr) {
	px := (*float64)(unsafe.Pointer(x))
	py := (*float64)(unsafe.Pointer(y))
	*px = math.Mod(*px, *py)
}

func go_modf(f, iptr uintptr) {
	pf := (*float64)(unsafe.Pointer(f))
	pint := (*float64)(unsafe.Pointer(iptr))
	*pint, *pf = math.Modf(*pf)
}

func go_atan2(y, x uintptr) {
	px := (*float64)(unsafe.Pointer(x))
	py := (*float64)(unsafe.Pointer(y))
	*py = math.Atan2(*py, *px)
}

func go_frexp(numptr, expptr uintptr) {
	pnum := (*float64)(unsafe.Pointer(numptr))
	pexp := (*int)(unsafe.Pointer(expptr)) //FIXME: is int OK here?
	*pnum, *pexp = math.Frexp(*pnum)
}

func go_ldexp(xptr uintptr, exp int) {
	px := (*float64)(unsafe.Pointer(xptr))
	*px = math.Ldexp(*px, exp)
}

func go_rand(retptr uintptr) {
	pret := (*int32)(unsafe.Pointer(retptr))
	*pret = rand.Int31()
}

func go_srand(seed uintptr) {
	rand.Seed(int64(seed))
}

func go_time(t uintptr) {
	pt := (*int64)(unsafe.Pointer(t))
	*pt = time.Now().Unix()
}

func golua_call(L, f, uv uintptr)

func poke(addr uintptr, value byte) {
	*(*byte)(unsafe.Pointer(addr)) = value
}

func pokestringplus0(dst uintptr, src []byte) {
	for i := 0; i < len(src); i++ {
		poke(dst+uintptr(i), src[i])
	}
	poke(dst+uintptr(len(src)), 0)
}

func golua_ctestrun(f uintptr)

func go_ctest(t_, f_ uintptr) {
	t := (*testing.T)(unsafe.Pointer(t_))
	defer func() {
		err := recover()
		if err == nil {
			return
		}
		where := "???"
		// skip panic() and panicstring()
		_, file, line, ok := runtime.Caller(3)
		if ok {
			where = fmt.Sprintf("%s:%d", file, line)
		}
		t.Fatalf("%s\nat %s", err, where)
	}()
	golua_ctestrun(f_)
}

type stod_state int

const (
	stod_mantissa stod_state = iota
	stod_afterdecimal
	stod_aftere
	stod_afteresign
)

func go_strtod(str, endptr_, hugeval_, mhugeval_, presult_ uintptr) {
	str0 := str
	for *peek(str) == ' ' { //FIXME: isWhitespace(...)
		str++
	}
	buf := make([]byte, 0)
	b := *peek(str)
	if b == '+' || b == '-' {
		buf = append(buf, b)
		str++
		b = *peek(str)
	}
	/*
	   point := false
	   e := false
	   esign := false
	*/
	state := stod_mantissa
	for {
		if '0' <= b && b <= '9' {
			if state == stod_aftere {
				state = stod_afteresign
			}
		} else if state == stod_mantissa && b == '.' {
			state = stod_afterdecimal
		} else if state <= stod_afterdecimal && (b == 'e' || b == 'E') {
			state = stod_aftere
		} else if state == stod_aftere && (b == '+' || b == '-') {
			state = stod_afteresign
		} else {
			break
		}
		/*
		   if '0'<=b && b<='9' {
		       if e && !esign {
		           esign=true
		       }
		   } else if !point && b=='.' {
		       point=true
		   } else if !e && (b=='e' || b=='E') {
		       e = true
		       point = true // decimal point not accepted in exponential part
		   } else if e && !esign && (b=='+' || b=='-') {
		       esign = true
		   } else {
		       break
		   }
		*/
		buf = append(buf, b)
		str++
		b = *peek(str)
	}

	result, err := strconv.ParseFloat(string(buf), 64)
	if err != nil {
		nerr, _ := err.(*strconv.NumError)
		isrange := nerr != nil && nerr.Err == strconv.ErrRange
		switch {
		case isrange && math.IsInf(result, 1):
			result = *(*float64)(unsafe.Pointer(hugeval_))
		case isrange && math.IsInf(result, -1):
			result = *(*float64)(unsafe.Pointer(mhugeval_))
		default:
			// TODO: for library testing, log the error somewhere
			result = 0
			str = str0
		}
	}
	*(*float64)(unsafe.Pointer(presult_)) = result

	if endptr_ != 0 {
		*(*uintptr)(unsafe.Pointer(endptr_)) = str
	}
}

type stdiostream int

var stdiostreams = make(map[stdiostream]*os.File)

//var stdiostreams map[stdiostream]*os.File
func init() {
	// see: golua_cstdlib.c, ·initlibs()
	stdiostreams[3] = os.Stdin
	stdiostreams[1] = os.Stdout
	stdiostreams[2] = os.Stderr
}

func (s stdiostream) putc(c byte) {
	// FIXME: check and return any errors from Write
	stdiostreams[s].Write([]byte{c})
}

func (s stdiostream) gets(n int) []byte {
	// FIXME: check and return any errors from Read
	buf := make([]byte, n)
	n, err := stdiostreams[s].Read(buf)
	if err != nil {
		panic(err.Error())
	} // FIXME: make sure errors are handled somehow sensibly
	return buf[:n]
}

func goputc(c_, stream_ uintptr) {
	stream := stdiostream(stream_)
	c := byte(c_)
	stream.putc(c)
}

func gofgets(s_ uintptr, n int, stream_ uintptr) {
	stream := stdiostream(stream_)
	buf := stream.gets(n - 1)
	pokestringplus0(s_, buf)
}
